precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile to interact with randomness through an evm precompile.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(assert_matches)]
+
+extern crate alloc;
+
+use fp_evm::{
+	Context, ExitReason, ExitSucceed, Log, Precompile, PrecompileHandle, PrecompileOutput,
+};
+use pallet_randomness::{BalanceOf, GetBabeData};
+use precompile_utils::{costs::call_cost, prelude::*};
+use sp_core::{H160, H256, U256};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+// #[cfg(test)]
+// mod mock;
+#[cfg(test)]
+mod tests;
+
+#[generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+	RelayBlockNumber = "relayBlockNumber()",
+	RelayEpochIndex = "relayEpochIndex()",
+	RequestBabeRandomnessCurrentBlock =
+		"requestBabeRandomnessCurrentBlock(address,uint256,uint64,bytes32,uint64)",
+	RequestBabeRandomnessOneEpochAgo =
+		"requestBabeRandomnessOneEpochAgo(address,uint256,uint64,bytes32)",
+	RequestBabeRandomnessTwoEpochsAgo =
+		"requestBabeRandomnessTwoEpochsAgo(address,uint256,uint64,bytes32)",
+	RequestLocalRandomness = "requestLocalRandomness(address,uint256,uint64,bytes32,uint64)",
+	FulfillRequest = "fulfillRequest(uint64)",
+	IncreaseRequestFee = "increaseRequestFee(uint64)",
+	ExecuteRequestExpiration = "executeRequestExpiration(uint64)",
+	InstantBabeRandomnessCurrentBlock = "instantBabeRandomnessCurrentBlock(uint64,bytes32)",
+	InstantBabeRandomnessOneEpochAgo = "instantBabeRandomnessOneEpochAgo(uint64,bytes32)",
+	InstantBabeRandomnessTwoEpochsAgo = "instantBabeRandomnessTwoEpochsAgo(uint64,bytes32)",
+	InstantLocalRandomness = "instantLocalRandomness(uint64,bytes32)",
+}
+
+pub const FULFILLMENT_ESTIMATED_COST: u64 = 1000u64; // TODO: get real value from benchmarking
+pub const LOG_SUBCALL_SUCCEEDED: [u8; 32] = keccak256!("SubcallSucceeded");
+pub const LOG_SUBCALL_FAILED: [u8; 32] = keccak256!("SubcallFailed");
+
+pub fn log_subcall_succeeded(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_SUCCEEDED)
+}
+
+pub fn log_subcall_failed(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_FAILED)
+}
+
+/// Reverts if fees and gas_limit are not sufficient to make subcall and cleanup
+fn ensure_can_provide_randomness<Runtime>(
+	code_address: H160,
+	gas_limit: u64,
+	request_fee: Option<BalanceOf<Runtime>>,
+	clean_up_cost: Option<u64>,
+) -> EvmResult<()>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	BalanceOf<Runtime>: Into<U256>,
+{
+	if let Some(fee) = request_fee {
+		// assert fee / base_fee > gasLimit
```

Comment:
You could instead do `fee > base_fee * gas_limit`, which avoids any convertion to smaller types.